components/core/tests/test-ir_serializer.cpp

```diff
@@ -0,0 +1,147 @@
+        four_byte_encoded_variable_t
+) {
+    LogEventSerializer<TestType> serializer;
+
+    // Test encoding with serializer
+    vector<string> var_strs
+            = {"4938",
+               std::to_string(INT32_MAX),
+               std::to_string(INT64_MAX),
+               "0.1",
+               "-25.519686",
+               "-25.5196868642755",
+               "-00.00",
+               "bin/python2.7.3",
+               "abc123"};
+    size_t var_ix{0};
+
+    string first_log_event = "here is first string with a small int " + var_strs[var_ix++];
+    auto const first_ts
+            = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+    first_log_event += " and a medium int " + var_strs[var_ix++];
+    first_log_event += " and a very large int " + var_strs[var_ix++];
+    first_log_event += " and a small float " + var_strs[var_ix++];
+    first_log_event += "\n";
+
+    auto second_log_event = "here is second string with a medium float " + var_strs[var_ix++];
+    auto const second_ts
+            = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+    second_log_event += " and a weird float " + var_strs[var_ix++];
+    second_log_event += " and a string with numbers " + var_strs[var_ix++];
+    second_log_event += " and another string with numbers " + var_strs[var_ix++];
+    second_log_event += "\n";
+
+    string ir_test_file = "ir_serializer_test";
+    ir_test_file += cIrFileExtension;
+
+    REQUIRE(serializer.open(ir_test_file));
+    // Test serializing log events
+    REQUIRE(serializer.serialize_log_event(first_ts, first_log_event));
+    REQUIRE(serializer.serialize_log_event(second_ts, second_log_event));
+    serializer.close();
+
+    Decompressor ir_reader;
+    ir_reader.open(ir_test_file);
+
+    bool uses_four_byte_encoding{false};
+    REQUIRE(IRErrorCode_Success
+            == clp::ffi::ir_stream::get_encoding_type(ir_reader, uses_four_byte_encoding));
+    REQUIRE(true == uses_four_byte_encoding);
+
+    auto result = LogEventDeserializer<TestType>::create(ir_reader);
+    REQUIRE(false == result.has_error());
+    auto deserializer_inst = std::move(result.value());
+
+    string decoded_message{};
+    // Deserialize the first log event from the IR
+    auto deserialized_result = deserializer_inst.deserialize_log_event();
+    REQUIRE(false == deserialized_result.has_error());
+
+    // Deserialize the log event
+    auto log_event = deserialized_result.value();
+    REQUIRE(clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success
+            == deserialize_log_event(log_event, decoded_message));
+
+    // Compare decoded message and timestamp
+    REQUIRE(decoded_message == first_log_event);
+    REQUIRE(log_event.get_timestamp() == first_ts);
+
+    // Deserialize the second log event from the IR
+    deserialized_result = deserializer_inst.deserialize_log_event();
+    REQUIRE(false == deserialized_result.has_error());
```

Comment:
Can we make reference log events into a vector, so that we can de-duplicate the deserialization part by looping?